Origin: http://selenic.com/hg/rev/8b7cd9a998f0
Description: copies: re-include root directory in directory rename detection (issue3511)
Bug: http://bugs.debian.org/698634
Bug-mercurial: http://bz.selenic.com/show_bug.cgi?id=3511
Applied-Upstream: 2.2.3

--- a/mercurial/context.py
+++ b/mercurial/context.py
@@ -1043,7 +1043,7 @@ class workingctx(changectx):
                 wlock.release()
 
     def dirs(self):
-        return self._repo.dirstate.dirs()
+        return set(self._repo.dirstate.dirs())
 
 class workingfilectx(filectx):
     """A workingfilectx object makes access to data related to a particular
--- a/mercurial/copies.py
+++ b/mercurial/copies.py
@@ -308,7 +308,9 @@ def mergecopies(repo, c1, c2, ca):
 
     # generate a directory move map
     d1, d2 = c1.dirs(), c2.dirs()
-    invalid = set([""])
+    d1.add('')
+    d2.add('')
+    invalid = set()
     dirmove = {}
 
     # examine each file copy for a potential directory move, which is
